src/node/desktop/src/main/desktop-browser-window.ts

```diff
@@ -276,6 +253,85 @@ export class DesktopBrowserWindow extends EventEmitter {
     });
   }
 
+  // implementation of DesktopWebPage.cpp::acceptNavigationRequest;
+  allowNavigation(url: string): boolean {
+    logger().logDebug(`allowNavigation checking ${url}`);
+
+    let targetUrl: URL;
+    try {
+      targetUrl = new URL(url);
+    } catch (err: unknown) {
+      // malformed URL will cause exception
+      logger().logError(err);
+      return false;
+    }
+
+    if (isAboutUrl(url) || isChromeGpuUrl(url)) {
+      return true;
+    }
+
+    if (!isAllowedProtocol(targetUrl)) {
+      logger().logDebug(`allowNavigation: disallowed protocol ${url}`);
+      return false;
+    }
+    
+    // determine if this is a local request (handle internally only if local)
+    const isLocal = isLocalUrl(targetUrl);
+
+    // iframe navigation interception: https://github.com/electron/electron/issues/7097
+    // need 'will-frame-navigate' event
+    // event handler should allow if the url is not local _and_ on the safe hosts list
+    // see DesktopWebPage.cpp::acceptNavigationRequest where it checks s_subframes
+    // workaround in GWT by calling DesktopFrame::allowNavigation to determine setting frame url
+
+    const base = this.options.baseUrl ?? this.mainWindow?.options.baseUrl;
+    if ((!base && isLocal)) {
+      return true;
+    }
+
+    if (base) {
+      const baseUrl = new URL(base);
+
+      if (targetUrl.protocol === baseUrl.protocol &&
+          targetUrl.host === baseUrl.host) {
+        return true;
+      }
+    }
+
+    const viewer = this.viewerUrl ?? this.mainWindow?.viewerUrl;
+    const tutorial = this.tutorialUrl ?? this.mainWindow?.tutorialUrl;
+    const presentation = this.presentationUrl ?? this.mainWindow?.presentationUrl;
+    const shinyDialog = this.shinyDialogUrl ?? this.mainWindow?.shinyDialogUrl;
```

Comment:
are there really any differences between these?

Reply:
Not really other than GWT has different callbacks for these special URL's. I think it does need to be tracked individually though since the different panes could all be displaying something. It would have been nice to refactor this so that the iframe (or something that represents it) keeps track of the allowed URL it can display.

Most of this was getting an understanding of what the URL navigation request was supposed to do and the URL's that are allowed.